jerry-core/api/jerry-module.c

```diff
@@ -0,0 +1,249 @@
+                      const jerry_value_t referrer, /**< parent module */
+                      void *user_p) /**< user data */
+{
+#if JERRY_MODULE_SYSTEM
+  JERRY_UNUSED (user_p);
+
+  const jerry_char_t *directory_p = lit_get_magic_string_utf8 (LIT_MAGIC_STRING__EMPTY);
+  jerry_size_t directory_size = lit_get_magic_string_size (LIT_MAGIC_STRING__EMPTY);
+
+  jerry_module_t *module_p = jerry_object_get_native_ptr (referrer, &jerry_module_native_info);
+
+  if (module_p != NULL)
+  {
+    directory_p = module_p->path_p;
+    directory_size = module_p->basename_offset;
+  }
+
+  jerry_size_t specifier_size = jerry_string_size (specifier, JERRY_ENCODING_UTF8);
+  jerry_size_t reference_size = directory_size + specifier_size;
+  jerry_char_t *reference_path_p = jerry_heap_alloc (reference_size + 1);
+
+  memcpy (reference_path_p, directory_p, directory_size);
+  jerry_string_to_buffer (specifier, JERRY_ENCODING_UTF8, reference_path_p + directory_size, specifier_size);
+  reference_path_p[reference_size] = '\0';
+
+  jerry_char_t *path_p = jerry_port_path_normalize (reference_path_p, reference_size);
+  jerry_heap_free (reference_path_p, reference_size + 1);
+
+  if (path_p == NULL)
+  {
+    return jerry_throw_sz (JERRY_ERROR_SYNTAX, "Error while accessing referenced module file");
+  }
+
+  jerry_value_t realm = jerry_current_realm ();
+
+  jerry_module_manager_t *manager_p;
+  manager_p = (jerry_module_manager_t *) jerry_context_data (&jerry_module_manager);
+
+  module_p = manager_p->module_head_p;
+
+  while (module_p != NULL)
+  {
+    if (module_p->realm == realm && strcmp ((const char *) module_p->path_p, (const char *) path_p) == 0)
+    {
+      jerry_value_free (realm);
+      jerry_port_path_free (path_p);
+      return jerry_value_copy (module_p->module);
+    }
+
+    module_p = module_p->next_p;
+  }
+
+  jerry_size_t source_size;
+  jerry_char_t *source_p = jerry_port_source_read ((const char *) path_p, &source_size);
+
+  if (source_p == NULL)
+  {
+    jerry_log (JERRY_LOG_LEVEL_ERROR, "Failed to open module file: %s\n", path_p);
+
+    jerry_value_free (realm);
+    jerry_port_path_free (path_p);
+
+    return jerry_throw_sz (JERRY_ERROR_SYNTAX, "Module file not found");
+  }
+
+  jerry_parse_options_t parse_options;
+  parse_options.options = JERRY_PARSE_MODULE | JERRY_PARSE_HAS_SOURCE_NAME;
+  parse_options.source_name = jerry_value_copy (specifier);
+
+  jerry_value_t ret_value = jerry_parse (source_p, source_size, &parse_options);
+  jerry_value_free (parse_options.source_name);
+
+  jerry_port_source_free (source_p);
+
+  if (jerry_value_is_exception (ret_value))
+  {
+    jerry_port_path_free (path_p);
+    jerry_value_free (realm);
+    return ret_value;
+  }
+
+  module_p = (jerry_module_t *) jerry_heap_alloc (sizeof (jerry_module_t));
+
+  module_p->next_p = manager_p->module_head_p;
+  module_p->path_p = path_p;
+  module_p->basename_offset = jerry_port_path_base (module_p->path_p);
```

Comment:
Maybe it's better add length parameter to jerry_port_path_base, there is no need strlen(module_p->path_p) again and again.

Reply:
We don't know the length of the string where the function is called, and we don't necessarly want to either. 
The port may not need the length at all, as it may just pass the string directly to a platform specific api, so calculating it in the core wouldn't be worth it. If the port implemntation needs the length, then it can calculate it as necessary.